### ISSUE TITLE
ownership not respected

### ISSUE THREAD

**Issue by natostanco** [docker-slim/docker-slim#26]
because docker copy does not preserve file owners, copying the whole slimmed root folder over changes the ownership of all the files to root. Software relying on proper ownership will have problems, some apps have inbuilt ownership fixers, some don't.

**Comment by kcq** [comment]
Yes, this definitely needs to be addressed. Let me know if you have one of those ownership sensitive applications. I'd love to hear about the use cases!

**Comment by sasikumar-swaminathan** [comment]
I have such a scenario where I trying to slim down the custom image built on top of CentOS with Pivotol TC server installed. All operations on TC server-specific files are to be performed by the user:group as `tcserver:tcserver`. However, the slimmed down image does not preserve them and change them to `root:0` which is not ideal.
Is there any workaround for this?

**Comment by callumgare** [comment]
I'd also find a solution for this very useful. My use case is that during the build process I set up a user to use for running my app with de-escalated privileges then set all copied in files to be readable by that user.

**Comment by pugnascotia** [comment]
Same issue here. My image uses ownership `1000:0`, which is changed to `0:0` in the slim version. I also noticed that `/tmp/` has group write removed, which means my app can't create a dir under `/tmp`.

I'm just experimenting at the moment, so I changed the slim `Dockerfile` to run some `chmod` and `chown` commands, but that means including those binaries in the image.

**Comment by kcq** [comment]
@pugnascotia which version of docker-slim are you using?

**Comment by kcq** [comment]
@callumgare are you using the `USER` instruction in your Dockerfile?

**Comment by callumgare** [comment]
> @callumgare are you using the `USER` instruction in your Dockerfile?

I’m no longer using docker-slim So I don’t remember exactly what I was doing but I believe so yes. I understand that it is discouraged at the moment but that was an important feature for me.

**Comment by kcq** [comment]
@callumgare Thanks for the clarification! Using the `USER` instruction is definitely not discouraged :-) There's a couple of constraints in the current implementation, but those will be addressed to take care of the side effects.

**Comment by pugnascotia** [comment]
@kcq I downloaded the 1.26.1 just the other day.

You've found the correct Dockerfile - I recently changed it to stop running as root. Elasticsearch will actually refuse to run as root, which makes things interesting :-) But it'd be awesome to get docker-slim working on our image. Let me know if I can help here.

**Comment by pugnascotia** [comment]
BTW, I've done a tiny bit of Go in the past, so I'm happy to roll my sleeves up.

**Comment by kcq** [comment]
@pugnascotia that'll be great if you can help out! Any help is appreciated including just reporting stuff like this... No need to be a Go expert :-) I'll be happy to do a code walk through to get you started. I'm pretty sure we'll be able to have a working ElasticSearch container and with the permission you want/need.

**Comment by kcq** [comment]
@pugnascotia silly question... what's the right way to build your docker image? That Dockerfile looks a bit strange on line 28 where it has `${source_elasticsearch}` and it's something `docker build` is not happy about :-)

**Comment by pugnascotia** [comment]
Ah yeah, there's some templating going on. You'll need to do e.g.

    ./gradlew distribution:docker:buildDockerImage

...and then you'll have an `elasticsearch:test` image to play with. There's more information in the `CONTRIBUTING.md` file in the repo, but just ping me if you have any issues.

**Comment by kcq** [comment]
Thanks for the pointers @pugnascotia ! It's a big repo with lots of stuff, so I wanted to ask first :)

**Comment by arturli** [comment]
Hey, did somebody bypass the problem with "ownership not respected" which @pugnascotia had ?
My app needs the normal settings for `/tmp`: 1777 - currently after running docker-slim (version: 1.26.1) it is as follows:
`sh-4.2$ ls -ld /tmp/                                                                                                                                                                                                                         
drwxr-xr-x. 8 root 0 4096 Feb 13 14:26 /tmp/
`

Thanks!

**Comment by kcq** [comment]
@arturli 1.26.2 will help with these use cases where you need specific permissions in your new images. It'll be out pretty soon... `docker-slim` will try to keep the original permissions and there'll be an extra/explicit flag to overwrite the permissions if/when you need to...

**Comment by kcq** [comment]
@pugnascotia The new release is out (1.27.0) and it does a better job with the ElasticSearch container image :) The scripts I used for the repro are here: https://github.com/docker-slim/examples/tree/master/3rdparty/elasticsearch

Note that I had to explicitly include the mmdb files used by the ingest-geoip plugin.

By the way, with the latest version you have an ability to overwrite the file permissions too.

**Comment by kcq** [comment]
@arturli Take a look at the new release (1.27.0)... The `/tmp` directory permissions are 177, by default, and you can change them to something else if you want using the `--include-path` and the `--path-perms`/`--path-perms-file` parameters.

**Comment by pugnascotia** [comment]
@kcq the latest release seems to be working well. I'll let you know if I run into any problems, but so far so good. Thank you 👍

**Comment by kcq** [comment]
@pugnascotia good to hear :) by the way, the next release will also add an extra flag to filter data in the included directories if necessary (using the `--exclude-pattern` parameter).
